### PR TITLE
Adds new Date Format String for HttpRuleParser (#507)

### PR DESCRIPTION
**Pull request by jkotalik**
@Tratcher

### CODE REVIEW

**Comment by Tratcher** [comment]
CLA required? @Eilon was it just the public one? https://github.com/aspnet/Home/blob/master/CONTRIBUTING.md#contributing-code-and-content

**Comment by Eilon** [comment]
I worked it out with @jkotalik so I think he'll be fine once he completes the process.

File: test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs [hunk 1]
```diff
@@ -16,6 +16,7 @@ public void TryParse_SetOfValidValueStrings_ParsedCorrectly()
             CheckValidParsedValue("Tue, 15 Nov 1994 08:12:31 GMT", new DateTimeOffset(1994, 11, 15, 8, 12, 31, TimeSpan.Zero));
             CheckValidParsedValue("      Sunday, 06-Nov-94 08:49:37 GMT   ", new DateTimeOffset(1994, 11, 6, 8, 49, 37, TimeSpan.Zero));
             CheckValidParsedValue(" Tue,\r\n 15 Nov\r\n 1994 08:12:31 GMT   ", new DateTimeOffset(1994, 11, 15, 8, 12, 31, TimeSpan.Zero));
+            CheckValidParsedValue("Sat, 09-Dec-2017 07:07:03 GMT ", new DateTimeOffset(2017, 12, 09, 7, 7, 3, TimeSpan.Zero));
```
Comment (Eilon) [inline test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs]: Should change this to a `Theory` test - one for each case.

File: test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs [hunk 2]
```diff
@@ -2,29 +2,41 @@
...
-        public void TryParse_SetOfValidValueStrings_ParsedCorrectly()
+        [Theory, MemberData(nameof(ValidStringData))]
+        public void TryParse_SetOfValidValueStrings_ParsedCorrectly(string input, DateTimeOffset expected)
         {
             // We don't need to validate all possible date values, since they're already tested in HttpRuleParserTest.
             // Just make sure the parser calls HttpRuleParser methods correctly.
-            CheckValidParsedValue("Tue, 15 Nov 1994 08:12:31 GMT", new DateTimeOffset(1994, 11, 15, 8, 12, 31, TimeSpan.Zero));
-            CheckValidParsedValue("      Sunday, 06-Nov-94 08:49:37 GMT   ", new DateTimeOffset(1994, 11, 6, 8, 49, 37, TimeSpan.Zero));
-            CheckValidParsedValue(" Tue,\r\n 15 Nov\r\n 1994 08:12:31 GMT   ", new DateTimeOffset(1994, 11, 15, 8, 12, 31, TimeSpan.Zero));
+            CheckValidParsedValue(input, expected);
```
Comment (Tratcher) [inline test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs]: You don't still need CheckValidParsedValue and CheckInvalidParsedValue do you?

**Comment by jkotalik** [comment]
@Eilon 🆙 📅

File: test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs [hunk 3]
```diff
@@ -2,29 +2,41 @@
...
+using System.Collections.Generic;
 using Xunit;
 
 namespace Microsoft.Net.Http.Headers
 {
     public class DateParserTest
     {
-        [Fact]
-        public void TryParse_SetOfValidValueStrings_ParsedCorrectly()
+        [Theory, MemberData(nameof(ValidStringData))]
```
Comment (Eilon) [inline test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs]: Per my comment in some other PR, split the attributes into two lines, each like `[Foo]`.

File: test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs [hunk 4]
```diff
@@ -2,29 +2,41 @@
...
-        public void TryParse_SetOfInvalidValueStrings_ReturnsFalse()
+        private static IEnumerable<object[]> ValidStringData()
+        {
+            yield return new object[] { "Tue, 15 Nov 1994 08:12:31 GMT", new DateTimeOffset(1994, 11, 15, 8, 12, 31, TimeSpan.Zero) };
+            yield return new object[] { "      Sunday, 06-Nov-94 08:49:37 GMT   ", new DateTimeOffset(1994, 11, 6, 8, 49, 37, TimeSpan.Zero) };
+            yield return new object[] { " Tue,\r\n 15 Nov\r\n 1994 08:12:31 GMT   ", new DateTimeOffset(1994, 11, 15, 8, 12, 31, TimeSpan.Zero) };
+            yield return new object[] { "Sat, 09-Dec-2017 07:07:03 GMT ", new DateTimeOffset(2017, 12, 09, 7, 7, 3, TimeSpan.Zero) };
+        }
+
+        [Theory, MemberData(nameof(InvalidStringData))]
```
Comment (Eilon) [inline test/Microsoft.Net.Http.Headers.Tests/DateParserTest.cs]: And this one.

**Comment by jkotalik** [comment]
@Tratcher  Went ahead and refactored the helpers. Will wait for this build to succeed to merge in.

**Pull request merged**